// react-native-number-please/dist/src/NumberPlease.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const React = __importStar(require("react"));
const react_native_1 = require("react-native");
const lodash_find_1 = __importDefault(require("lodash.find"));
const lodash_findindex_1 = __importDefault(require("lodash.findindex"));
const immer_1 = __importDefault(require("immer"));
const range_1 = __importDefault(require("./utils/range"));
const PickerFactory = ({ pickerProps, selectedValue, onChange, pickerStyle, itemStyle, }) => {
    const { id, label = '', min, max } = pickerProps;
    const numbers = range_1.default(min, max);
    return (<react_native_1.Picker style={{ ...styles.picker, ...pickerStyle }} selectedValue={selectedValue} onValueChange={(value) => onChange({ id, value })} itemStyle={itemStyle}>
      {numbers.map((number, index) => (<react_native_1.Picker.Item key={`${id}-${number}-${index}`} value={number} label={`${number} ${label}`}/>))}
    </react_native_1.Picker>);
};
const NumberPlease = ({ digits, values, onChange, ...rest }) => {
    const onChangeHandle = (value) => {
        const nextValues = immer_1.default(values, (draft) => {
            const index = lodash_findindex_1.default(draft, { id: value.id });
            draft[index] = value;
        });
        onChange(nextValues);
    };
    return (<react_native_1.View style={styles.container}>
      {digits.map((picker, index) => {
        const pickerValues = lodash_find_1.default(values, { id: picker.id });
        return (<PickerFactory key={`${picker.id}-picker-${index}`} pickerProps={picker} selectedValue={pickerValues?.value} onChange={onChangeHandle} {...rest}/>);
    })}
    </react_native_1.View>);
};
const styles = react_native_1.StyleSheet.create({
    container: {
        flexDirection: 'row',
        width: '100%',
        justifyContent: 'center',
        alignItems: 'center',
    },
    picker: {
        height: '100%', width: 90,
    },
});
exports.default = NumberPlease;
