// json-to-table-pagination/dist/index.modern.js
import React, { useState, useEffect } from 'react';
import 'bootstrap/dist/css/bootstrap.min.css';
import { BsChevronDoubleLeft, BsChevronLeft, BsChevronRight, BsChevronDoubleRight } from 'react-icons/bs';
import styled from 'styled-components';

function _taggedTemplateLiteralLoose(strings, raw) {
  if (!raw) {
    raw = strings.slice(0);
  }

  strings.raw = raw;
  return strings;
}

var PRIMARY = '#1E88E5';
var FIRST = 'First';
var LAST = 'Last';
var PREVIOUS = 'Previous';
var NEXT = 'NEXT';

function _templateObject7() {
  var data = _taggedTemplateLiteralLoose(["\n    width: 6rem;\n"]);

  _templateObject7 = function _templateObject7() {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = _taggedTemplateLiteralLoose(["\n    align-items: center;\n    display: flex;\n    justify-content: flex-end;\n    padding-right: 1.3rem;\n"]);

  _templateObject6 = function _templateObject6() {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = _taggedTemplateLiteralLoose(["\n    list-style: none;\n    display: flex;\n    justify-content: center;\n    margin-bottom: 0;\n\n    li {\n        align-items: center;\n        border: 1px solid white;\n        cursor: pointer;\n        display: flex;\n        font-weight: bold;\n        margin-right: 2rem;\n        padding: 0.5rem;\n    }\n\n    li:third-child {\n        cursor: none;\n    }\n"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = _taggedTemplateLiteralLoose(["\n    display: flex;\n    justify-content: space-around;\n    padding: 0.5rem;\n    border-radius: 0;\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = _taggedTemplateLiteralLoose(["\n    width: 5rem;\n    padding: 0.5rem;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteralLoose(["\n    align-items: center;\n    display: flex;\n    padding-left: 1.3rem;\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteralLoose(["\n    background-color: ", ";\n    color: white;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}
var PaginationWrapper = styled.div(_templateObject(), function (props) {
  return props.bgColor ? props.bgColor : PRIMARY;
});
var PaginationRowsWrapper = styled.div(_templateObject2());
var PaginationSelect = styled.select(_templateObject3());
var PaginationNav = styled.nav(_templateObject4());
var PaginationList = styled.ul(_templateObject5());
var PageNumberWrapper = styled.div(_templateObject6());
var PaginationNumberInput = styled.input(_templateObject7());

var Pagination = (function (_ref) {
  var currentPage = _ref.currentPage,
      getNextPage = _ref.getNextPage,
      getPreviousPage = _ref.getPreviousPage,
      getFirstPage = _ref.getFirstPage,
      getLastPage = _ref.getLastPage,
      displayRowsPerPage = _ref.displayRowsPerPage,
      bgColor = _ref.bgColor,
      _ref$useArrowIcons = _ref.useArrowIcons,
      useArrowIcons = _ref$useArrowIcons === void 0 ? true : _ref$useArrowIcons,
      goToPage = _ref.goToPage,
      searchTerm = _ref.searchTerm;
  return /*#__PURE__*/React.createElement(PaginationWrapper, {
    className: "row",
    bgColor: bgColor
  }, /*#__PURE__*/React.createElement(PaginationRowsWrapper, {
    className: "col-md-1"
  }, /*#__PURE__*/React.createElement(PaginationSelect, {
    className: "pagination-select",
    onChange: displayRowsPerPage
  }, /*#__PURE__*/React.createElement("option", {
    value: "5"
  }, "5"), /*#__PURE__*/React.createElement("option", {
    value: "10"
  }, "10"), /*#__PURE__*/React.createElement("option", {
    value: "25"
  }, "25"), /*#__PURE__*/React.createElement("option", {
    value: "50"
  }, "50"), /*#__PURE__*/React.createElement("option", {
    value: "100"
  }, "100"))), /*#__PURE__*/React.createElement(PaginationNav, {
    className: "col-md-9"
  }, /*#__PURE__*/React.createElement(PaginationList, {
    className: "pagination-list"
  }, /*#__PURE__*/React.createElement("li", {
    onClick: function onClick() {
      return getFirstPage();
    }
  }, useArrowIcons ? /*#__PURE__*/React.createElement(BsChevronDoubleLeft, null) : FIRST), /*#__PURE__*/React.createElement("li", {
    onClick: function onClick() {
      return getPreviousPage();
    }
  }, useArrowIcons ? /*#__PURE__*/React.createElement(BsChevronLeft, null) : PREVIOUS), /*#__PURE__*/React.createElement("li", {
    className: "current-page"
  }, currentPage), /*#__PURE__*/React.createElement("li", {
    onClick: function onClick() {
      return getNextPage();
    }
  }, useArrowIcons ? /*#__PURE__*/React.createElement(BsChevronRight, null) : NEXT), /*#__PURE__*/React.createElement("li", {
    onClick: function onClick() {
      return getLastPage();
    }
  }, useArrowIcons ? /*#__PURE__*/React.createElement(BsChevronDoubleRight, null) : LAST))), /*#__PURE__*/React.createElement(PageNumberWrapper, {
    className: "col-md-2"
  }, /*#__PURE__*/React.createElement(PaginationNumberInput, {
    type: "text",
    placeholder: "Go to page",
    onChange: goToPage,
    value: searchTerm,
    className: "pagination-input"
  })));
});

function _templateObject3$1() {
  var data = _taggedTemplateLiteralLoose(["\n    user-select: all;\n    word-break: break-word;\n    width: 50px;\n"]);

  _templateObject3$1 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2$1() {
  var data = _taggedTemplateLiteralLoose(["\n    border: 1px solid #fafafa;\n"]);

  _templateObject2$1 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject$1() {
  var data = _taggedTemplateLiteralLoose(["\n    text-transform: capitalize;\n    background-color: ", ";\n    color: white;\n\n    th {\n        cursor: pointer;\n    }\n"]);

  _templateObject$1 = function _templateObject() {
    return data;
  };

  return data;
}
var TableHead = styled.thead(_templateObject$1(), function (props) {
  return props.bgColor ? props.bgColor : PRIMARY;
});
var CustomTable = styled.div(_templateObject2$1());
var TableData = styled.td(_templateObject3$1());

var sortArrayObjects = function sortArrayObjects(key, order) {
  if (order === void 0) {
    order = true;
  }

  return function innerSort(a, b) {
    if (!a.hasOwnProperty(key) || !b.hasOwnProperty(key)) {
      return 0;
    }

    var varA = typeof a[key] === 'string' ? a[key].toUpperCase() : a[key];
    var varB = typeof b[key] === 'string' ? b[key].toUpperCase() : b[key];
    var comparison = 0;

    if (varA > varB) {
      comparison = 1;
    } else if (varA < varB) {
      comparison = -1;
    }

    return order === false ? comparison * -1 : comparison;
  };
};
var getHeaders = function getHeaders(jsonFile) {
  var headers = [];
  jsonFile && jsonFile.map(function (obj) {
    var objKeys = Object.keys(obj);
    objKeys.map(function (key) {
      if (!headers.includes(key)) {
        headers.push(key);
      }
    });
  });
  return headers;
};
var flattenNestedObjects = function flattenNestedObjects(jsonFile) {
  return jsonFile.map(function (obj) {
    return flattenObjectFunc(obj);
  });
};

var flattenObjectFunc = function flattenObjectFunc(ob) {
  var toReturn = {};

  for (var i in ob) {
    if (!ob.hasOwnProperty(i)) continue;

    if (typeof ob[i] == 'object') {
      var flatObject = flattenObjectFunc(ob[i]);

      for (var x in flatObject) {
        if (!flatObject.hasOwnProperty(x)) continue;
        toReturn[i + '.' + x] = flatObject[x];
      }
    } else {
      toReturn[i] = ob[i];
    }
  }

  return toReturn;
};

var CustomTable$1 = (function (_ref) {
  var jsonFile = _ref.jsonFile,
      options = _ref.options;
  var bgColor = options.bgColor,
      useArrowIcons = options.useArrowIcons,
      flattenObjects = options.flattenObjects;

  var _useState = useState(1),
      currentPage = _useState[0],
      setCurrentPage = _useState[1];

  var _useState2 = useState(5),
      rowsPerPage = _useState2[0],
      setRowsPerPage = _useState2[1];

  var _useState3 = useState(''),
      searchTerm = _useState3[0],
      setSearchTerm = _useState3[1];

  var _useState4 = useState(true),
      sortOrder = _useState4[0],
      setSortOrder = _useState4[1];

  var _useState5 = useState(jsonFile),
      filteredRecords = _useState5[0],
      setFilteredRecords = _useState5[1];

  var _useState6 = useState([]),
      tableHeaders = _useState6[0],
      setTableHeaders = _useState6[1];

  useEffect(function () {
    if (flattenObjects) {
      var flattenedObjects = flattenNestedObjects(jsonFile);
      setFilteredRecords(flattenedObjects);
      handleTableHeaders(flattenedObjects);
    } else {
      handleTableHeaders(jsonFile);
    }
  }, []);

  var handleTableHeaders = function handleTableHeaders(file) {
    var headers = getHeaders(file);
    setTableHeaders(headers);
  };

  var totalPages = Math.ceil(filteredRecords.length / rowsPerPage);
  var indexOfLastRow = currentPage * rowsPerPage;
  var indexOfFirstRow = indexOfLastRow - rowsPerPage;
  var currentRecords = filteredRecords.slice(indexOfFirstRow, indexOfLastRow);

  var renderTableRows = function renderTableRows() {
    return currentRecords.map(function (currentRecord, index) {
      return /*#__PURE__*/React.createElement("tr", {
        key: index.toString()
      }, tableHeaders.map(function (header, indexKey) {
        var value = currentRecord[header];
        return /*#__PURE__*/React.createElement(TableData, {
          key: indexKey.toString()
        }, typeof value === 'object' ? JSON.stringify(value) : value);
      }));
    });
  };

  var handleSortColumn = function handleSortColumn(header) {
    setSortOrder(!sortOrder);
    setFilteredRecords(filteredRecords.sort(sortArrayObjects(header, sortOrder)));
  };

  var renderTableHeader = function renderTableHeader() {
    return tableHeaders.map(function (header, index) {
      return /*#__PURE__*/React.createElement("th", {
        key: index.toString(),
        onClick: function onClick() {
          return handleSortColumn(header);
        }
      }, header);
    });
  };

  var getNextPage = function getNextPage() {
    if (currentPage !== totalPages) {
      setCurrentPage(currentPage + 1);
    }
  };

  var getPreviousPage = function getPreviousPage() {
    if (currentPage !== 1) {
      setCurrentPage(currentPage - 1);
    }
  };

  var getFirstPage = function getFirstPage() {
    return setCurrentPage(1);
  };

  var getLastPage = function getLastPage() {
    return setCurrentPage(totalPages);
  };

  var displayRowsPerPage = function displayRowsPerPage(e) {
    setRowsPerPage(e.target.value);
    setCurrentPage(1);
  };

  var goToPage = function goToPage(e) {
    var pageNumber = e.target.value;

    if (pageNumber > 0 && pageNumber <= totalPages) {
      setCurrentPage(pageNumber);
      setSearchTerm(pageNumber);
    }

    if (pageNumber === '') {
      setSearchTerm('');
      setCurrentPage(1);
    }
  };

  return /*#__PURE__*/React.createElement(CustomTable, null, /*#__PURE__*/React.createElement("table", {
    className: "table table-hover table-striped"
  }, /*#__PURE__*/React.createElement(TableHead, {
    bgColor: bgColor
  }, /*#__PURE__*/React.createElement("tr", null, renderTableHeader())), /*#__PURE__*/React.createElement("tbody", null, renderTableRows())), /*#__PURE__*/React.createElement(Pagination, {
    currentPage: currentPage,
    getNextPage: getNextPage,
    getPreviousPage: getPreviousPage,
    getFirstPage: getFirstPage,
    getLastPage: getLastPage,
    displayRowsPerPage: displayRowsPerPage,
    bgColor: bgColor,
    useArrowIcons: useArrowIcons,
    goToPage: goToPage,
    searchTerm: searchTerm
  }));
});

var JSONToTable = function JSONToTable(_ref) {
  var jsonFile = _ref.jsonFile,
      options = _ref.options;
  return /*#__PURE__*/React.createElement(CustomTable$1, {
    jsonFile: jsonFile,
    options: options
  });
};

export { JSONToTable };
